google/cloud/internal/external_account_integration_test.cc

```diff
@@ -38,22 +43,56 @@ TEST(ExternalAccountIntegrationTest, UrlSourced) {
   // TODO(#5915) - use higher-level abstractions once available
   auto json = nlohmann::json::parse(contents, nullptr, false);
   ASSERT_TRUE(json.is_object()) << "json=" << json.dump();
+
+  auto ec = internal::ErrorContext(
+      {{"GOOGLE_CLOUD_CPP_EXTERNAL_ACCOUNT_FILE", *filename},
+       {"program", "test"}});
+  auto type = ValidateStringField(json, "type", "credentials-file", ec);
+  ASSERT_STATUS_OK(type);
+  ASSERT_EQ(*type, "external_account");
+
+  auto audience = ValidateStringField(json, "audience", "credentials-file", ec);
+  ASSERT_STATUS_OK(audience);
+  auto subject_token_type =
+      ValidateStringField(json, "subject_token_type", "credentials-file", ec);
+  ASSERT_STATUS_OK(subject_token_type);
+  auto token_url =
+      ValidateStringField(json, "token_url", "credentials-file", ec);
+  ASSERT_STATUS_OK(token_url);
+
   ASSERT_TRUE(json.contains("credential_source")) << "json=" << json.dump();
   auto credential_source = json["credential_source"];
   ASSERT_TRUE(credential_source.is_object()) << "json=" << json.dump();
-  auto make_client = []() {
-    return rest_internal::MakeDefaultRestClient("", Options{});
+
+  auto make_client = [](Options opts = {}) {
+    return rest_internal::MakeDefaultRestClient("", std::move(opts));
   };
-  auto source = MakeExternalAccountTokenSourceUrl(
-      credential_source, make_client,
-      internal::ErrorContext(
-          {{"GOOGLE_CLOUD_CPP_EXTERNAL_ACCOUNT_FILE", *filename},
-           {"program", "test"}}));
+  auto source =
+      MakeExternalAccountTokenSourceUrl(credential_source, make_client, ec);
   ASSERT_STATUS_OK(source);
-  auto subject_token = (*source)(Options{});
-  ASSERT_STATUS_OK(subject_token);
-  std::cout << "subject_token=" << subject_token->token.substr(0, 32)
-            << "...<truncated>...\n";
+
+  auto info =
+      ExternalAccountInfo{*std::move(audience), *std::move(subject_token_type),
+                          *std::move(token_url), *std::move(source)};
+  auto credentials = ExternalAccountCredentials(info, make_client);
+  // Anything involving HTTP requests may fail and needs a retry loop.
```

Comment:
Forgive my ignorance of oauth2 business, but does our library perform some sort of retry loop like this? If not, must we?

Reply:
No. We rely on the main event loop.  I created #10345 because there is hope to make this better.